### PR TITLE
patch rep detection and outputs linking for complex top h5md

### PR DESCRIPTION
**Pull request by JFRudzinski**
### 1) `model_system_ref` carryover became robust for mismatched `model_system` vs `outputs`
- **File:** [outputs.py](/home/jfrudzinski/work/soft/nomad-distro-dev-topology-normalizer-2026-02/packages/nomad-simulations/src/nomad_simulations/schema_packages/outputs.py:235)
- **What changed:** `Outputs.set_model_system_ref()` no longer returns `None` when lengths differ.
- **New fallback order:**
  1. use 1:1 index mapping if lengths match
  2. use `representative_system_index` (supports both `int` and `np.integer`)
  3. use first `model_system` that has `particle_states`
  4. fallback to last `model_system`
- **Why:** downstream normalization (DOS/MD/topology-related mapping) depends on `output.model_system_ref`; previously this could be missing in trajectory-shaped archives.

### 2) MD bead-group selection no longer depends on fragile MDAnalysis selection strings
- **File:** [molecular_dynamics.py](/home/jfrudzinski/work/soft/nomad-distro-dev-topology-normalizer-2026-02/packages/nomad-simulations/src/nomad_simulations/schema_packages/utils/molecular_dynamics.py:767)
- **What changed:** `_get_molecular_bead_groups()` switched from:
  - `universe.select_atoms('moltype ' + moltype)`
  to:
  - boolean mask on `universe.atoms.moltypes` (`universe.atoms[atoms_moltypes == moltype]`)
- **Why:** avoids `MDAnalysis.exceptions.SelectionError` for plain tokens like `water`.

### 3) Tests updated/added
- **File:** [test_outputs.py](/home/jfrudzinski/work/soft/nomad-distro-dev-topology-normalizer-2026-02/packages/nomad-simulations/tests/test_outputs.py:98)
  - updated mismatch expectations for `set_model_system_ref()`
  - added `test_set_model_system_ref_prefers_representative_system_on_mismatch`
- **File:** [test_molecular_dynamics.py](/home/jfrudzinski/work/soft/nomad-distro-dev-topology-normalizer-2026-02/packages/nomad-simulations/tests/utils/test_molecular_dynamics.py:215)
  - added regression `test_get_molecular_bead_groups_handles_plain_moltype_tokens`

### CODE REVIEW

**Comment by JFRudzinski** [comment]
@ndaelman-hu This is a small patch to #373 -- if you haven't looked at that one yet, we can just merge it immediately. I just kept it separate for clarity in case. NOTE: This PR is to merge into the #373 branch

File: src/nomad_simulations/schema_packages/utils/molecular_dynamics.py [hunk 1]
```diff
@@ -764,8 +764,14 @@ def _get_molecular_bead_groups(
...
     for moltype in moltypes:
-        ags_by_moltype = universe.select_atoms('moltype ' + moltype)
+        try:
+            # Avoid text-based selection parsing (e.g., `moltype water`) which can
+            # fail for valid moltype values depending on tokenizer semantics.
+            ags_by_moltype = universe.atoms[atoms_moltypes == moltype]
+        except Exception:
+            ags_by_moltype = universe.atoms[:0]
         if ags_by_moltype.n_atoms == 0:
             continue
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/utils/molecular_dynamics.py]: In the try-except block: while it's nice to be defensive, I don't see any reason why this lookup would ever fail, though. Neither does the code, as it shields from a general `Exception`.
Moreover, this double pass on `atoms_moltypes` is unnecessary.


```suggestion
atoms_moltypes = np.asarray(getattr(universe.atoms, 'moltypes', []))

if not moltypes:
    moltypes = np.unique(atoms_moltypes)  # No .tolist()!

bead_groups = {}
for moltype in moltypes:
    ags_by_moltype = universe.atoms[atoms_moltypes == moltype]

    if ags_by_moltype.n_atoms == 0:
        continue

    if ags_by_moltype.masses is not None:
        ags_by_moltype = ags_by_moltype[ags_by_moltype.masses > abs(1e-2)]

    bead_groups[moltype] = BeadGroup(ags_by_moltype, compound='fragments')

return bead_groups
```

File: src/nomad_simulations/schema_packages/outputs.py [hunk 2]
```diff
@@ -242,13 +242,30 @@ def set_model_system_ref(self) -> ModelSystem | None:
...
             model_systems = self.m_parent.model_system
             outputs = self.m_parent.outputs
-            if (
-                isinstance(model_systems, list)
-                and isinstance(outputs, list)
-                and len(model_systems) == len(outputs)
-                and len(model_systems) > 0
-            ):
+            if not isinstance(model_systems, list) or len(model_systems) == 0:
+                return None
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/outputs.py]: This will fail for other repeating structures like `np.array`. I would just duck-type isntead:


```suggestion
            if model_systems is None or len(model_systems) == 0:
                return None
```

File: tests/utils/test_molecular_dynamics.py [hunk 3]
```diff
@@ -209,3 +210,18 @@ def test_archive_to_universe_base_model_method_no_attribute_error():
...
+    """
+    universe = MDAnalysis.Universe.empty(3, n_residues=3, atom_resindex=[0, 1, 2])
+    universe.add_TopologyAttr('moltypes', np.array(['water', 'water', 'water']))
+    universe.add_TopologyAttr('masses', np.array([15.999, 1.008, 1.008]))
+    universe.add_TopologyAttr('bonds', [(0, 1), (1, 2)])
+
+    bead_groups = _get_molecular_bead_groups(universe)
+
+    assert 'water' in bead_groups
+    assert len(bead_groups['water']) == 1
```
Comment (ndaelman-hu) [inline tests/utils/test_molecular_dynamics.py]: These are 2 tests merged into one. Would keep the objectives clear and separate them out:


```suggestion
def test_get_molecular_bead_groups_handles_reserved_keyword_moltypes():
    """Regression: moltype='water' should not raise `SelectionError`. The old text-based selection ('moltype water') failed because 'water' is  a reserved MDAnalysis keyword. Boolean indexing avoids the parser entirely and works with any moltype value, including reserved keywords."""
    universe = MDAnalysis.Universe.empty(1)
    universe.add_TopologyAttr('moltypes', np.array(['water']))

    bead_groups = _get_molecular_bead_groups(universe)

    assert 'water' in bead_groups


def test_get_molecular_bead_groups_counts_fragments():
    """BeadGroup counts bonded atoms as a single fragment."""
    universe = MDAnalysis.Universe.empty(3, n_residues=3, atom_resindex=[0, 1, 2])
    universe.add_TopologyAttr('moltypes', np.array(['mol', 'mol', 'mol']))
    universe.add_TopologyAttr('bonds', [(0, 1), (1, 2)])

    bead_groups = _get_molecular_bead_groups(universe)

    assert 'mol' in bead_groups
    assert len(bead_groups['mol']) == 1  # 3 bonded atoms = 1 fragment
```

File: tests/test_outputs.py [hunk 4]
```diff
@@ -118,9 +118,29 @@ def test_set_model_system_ref(
             if len(model_systems) == len(outputs_list) and len(model_systems) > 0:
                 assert model_system_ref == model_systems[i]
                 assert model_system_ref.name == f'system_{i}'
+            elif len(model_systems) > 0:
+                assert model_system_ref == model_systems[-1]
             else:
                 assert model_system_ref is None
 
+    def test_set_model_system_ref_prefers_representative_system_on_mismatch(self):
```
Comment (ndaelman-hu) [inline tests/test_outputs.py:126]: Pydocs


```suggestion
    def test_set_model_system_ref_prefers_representative_system_on_mismatch(self):
    """When model_system and outputs lengths mismatch, use representative_system_index. Tests the fallback chain: with 3 model_systems and 2 outputs, the 1:1 mapping fails, so set_model_system_ref() should use representative_system_index instead of falling back to last system or particle_states logic."""

```

**Review by ndaelman-hu — APPROVED** [review]
PR fixes bugs in the H5MD parsing and MD trajectory analysis.

Some minor code quality suggestions. Pls merge at your convenience, as the target branch isn't `develop` yet and further issues can be spotted down the line.

File: src/nomad_simulations/schema_packages/utils/molecular_dynamics.py (hunk 1 above)
Comment (JFRudzinski) [reply src/nomad_simulations/schema_packages/utils/molecular_dynamics.py]: Made a slight adjustment to your approach: Refactored _get_molecular_bead_groups to remove the duplicate atoms_moltypes pass and mutable default, while replacing the broad try/except with an explicit shape guard before boolean indexing to keep edge-case safety

File: src/nomad_simulations/schema_packages/outputs.py (hunk 2 above)
Comment (JFRudzinski) [reply src/nomad_simulations/schema_packages/outputs.py]: implemented as suggested

File: tests/utils/test_molecular_dynamics.py (hunk 3 above)
Comment (JFRudzinski) [reply tests/utils/test_molecular_dynamics.py]: implemented pretty much as is, just a slight addition to the keyword test

File: tests/test_outputs.py (hunk 4 above)
Comment (JFRudzinski) [reply tests/test_outputs.py:126]: nearly identical
